### PR TITLE
simplify installer

### PR DESCRIPTION
**Pull request by balamurugana**
Signed-off-by: Bala.FA <bala@minio.io>

### CODE REVIEW

File: pkg/installer/storageclass.go [hunk 1]
```diff
@@ -20,134 +20,154 @@ import (
...
-	return nil
-}
-
-func uninstallStorageClassDefault(ctx context.Context, c *Config) error {
-	if err := executeFn(ctx, c, "storage class", deleteStorageClassDefault); err != nil {
-		return fmt.Errorf("unable to delete storage class; %v", err)
+func createStorageClass(ctx context.Context, args *Args) error {
+	var gvk *schema.GroupVersionKind
+	if args.DryRun {
+		if args.KubeVersion.Major() >= 1 && args.KubeVersion.Minor() < 16 {
```
Comment (Praveenrajmani) [inline pkg/installer/storageclass.go:41]:
```suggestion
		if args.KubeVersion.Major() == 1 && args.KubeVersion.Minor() < 16 {
```

File: pkg/installer/installer.go [hunk 2]
```diff
@@ -18,100 +18,246 @@ package installer
...
+	err := args.validate()
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case args.DryRun:
+		if args.KubeVersion == nil {
+			// default higher version
+			if args.KubeVersion, err = version.ParseSemantic("1.15.0"); err != nil {
```
Comment (Praveenrajmani) [inline pkg/installer/installer.go:77]: can we chose a higher version incase if kubeVersion is not provided? `1.15.0` is very old and deprecated

Comment (Praveenrajmani) [reply pkg/installer/installer.go:77]: we can choose a latest version

Comment (Praveenrajmani) [reply pkg/installer/installer.go:77]: Or can we make `--kube-version` as mandatory for dry-run mode?

File: pkg/installer/deployment.go [hunk 3]
```diff
@@ -115,52 +101,303 @@ func createDeployment(ctx context.Context, c *Config) error {
...
+				},
+				Spec: podSpec,
+			},
+		},
+		Status: appsv1.DeploymentStatus{},
+	}
+
+	if args.DryRun {
+		fmt.Print(mustGetYAML(deployment))
+		return nil
```
Comment (Praveenrajmani) [inline pkg/installer/deployment.go:147]: In  dry-run mode, we should also print the admin-service

File: pkg/installer/args.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		return nil, errors.New("image name must be provided")
+	}
+
+	if credential == nil {
+		return nil, errors.New("credential must be provided")
+	}
+
+	if auditWriter == nil {
+		return nil, errors.New("audit writer must be provided")
+	}
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply pkg/installer/args.go:72]: Done

File: pkg/installer/crd.go [hunk 5]
```diff
@@ -18,47 +18,161 @@ package installer
...
+			return newCRD.Spec.Versions[i], nil
+		}
+	}
+
+	return crdVersion, fmt.Errorf("no version %v found crd %v", consts.LatestAPIVersion, newCRD.Name)
+}
+
+func updateCRD(
+	ctx context.Context,
+	args *Args,
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply pkg/installer/crd.go:81]: Done

File: pkg/installer/crd.go [hunk 6]
```diff
@@ -18,47 +18,161 @@ package installer
...
+	existingCRD, newCRD *apiextensions.CustomResourceDefinition,
+) (*apiextensions.CustomResourceDefinition, error) {
+	existingCRDStorageVersion, err := apihelpers.GetCRDStorageVersion(existingCRD)
+	if err != nil {
+		return nil, err
+	}
+
+	// CRD is already in the latest version
+	if existingCRDStorageVersion == consts.LatestAPIVersion {
+		return existingCRD, nil
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply pkg/installer/crd.go:91]: Done

File: pkg/installer/csidriver.go [hunk 7]
```diff
@@ -20,51 +20,48 @@ import (
...
-	return nil
-}
-
-func uninstallCSIDriverDefault(ctx context.Context, c *Config) error {
-	if err := executeFn(ctx, c, "CSI Driver", deleteCSIDriver); err != nil {
-		return fmt.Errorf("unable to delete CSI Driver; %v", err)
+func createCSIDriver(ctx context.Context, args *Args) error {
+	var gvk *schema.GroupVersionKind
+	if args.DryRun {
+		if args.KubeVersion.Major() >= 1 && args.KubeVersion.Minor() < 19 {
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply pkg/installer/csidriver.go:39]: No. Any non-standard kubernetes, we need to use lower version for safer side.

File: pkg/installer/storageclass.go (hunk 1 above)
Comment (balamurugana) [reply pkg/installer/storageclass.go:41]: No. Any non-standard kubernetes, we need to use lower version for safer side.

File: pkg/installer/installer.go (hunk 2 above)
Comment (balamurugana) [reply pkg/installer/installer.go:77]: Done. It was supposed to be `1.25.0`

File: pkg/installer/deployment.go (hunk 3 above)
Comment (balamurugana) [reply pkg/installer/deployment.go:147]: It is done at `createAdminService()`

Comment (Praveenrajmani) [reply pkg/installer/deployment.go:147]: when `--dry-run` is set, we return `nil` right away without printing the admin service

Comment (balamurugana) [reply pkg/installer/deployment.go:147]: Fixed `createAdminServerDeployment()`

**Review by Praveenrajmani — APPROVED** [review]

**Pull request merged**